src/gpl/src/initialPlace.cpp

```diff
@@ -106,23 +109,56 @@ void InitialPlace::doBicgstabPlace() {
 
   // set ExtId for idx reference // easy recovery
   setPlaceInstExtId();
-  for(int i=1; i<=ipVars_.maxIter; i++) {
-    updatePinInfo();
-    createSparseMatrix();
 
-    // BiCGSTAB solver for initial place
-    BiCGSTAB< SMatrix, IdentityPreconditioner > solver;
-    solver.setMaxIterations(ipVars_.maxSolverIter);
-    solver.compute(placeInstForceMatrixX_);
-    instLocVecX_ = solver.solveWithGuess(fixedInstForceVecX_, instLocVecX_);
-    errorX = solver.error();
+  for(size_t iter=1; iter<=ipVars_.maxIter; iter++) {
 
-    solver.compute(placeInstForceMatrixY_);
-    instLocVecY_ = solver.solveWithGuess(fixedInstForceVecY_, instLocVecY_);
-    errorY = solver.error();
+    updatePinInfo();
+    createSparseMatrix();
+    #if CUDA
+      // Set sparse matrices from dense matrices.
```

Comment:
You might want to break these two #ifs into two different methods so it looks more like this

#if CUDA
  cudaSparseSolve(...);
#else
 cpuSparseSolve(...)
#endif

Reply:
Is the current version okay or is it better to clear them more to one or two lines like "cuda**" and "cpu**"?
(I sent the screenshot to the chat)

Reply:
Include the screenshot here so we can all see

Reply:
![Screenshot 2022-07-27 3 52 15 PM](https://user-images.githubusercontent.com/57721574/181387047-48b5dc13-9761-405f-af16-d8c4359eecf2.png)

Reply:
Sorry I didn't know I can send pictures here.

Reply:
You'll have to get rid of the #if to make it runtime switchable (see my other feedback).  So I think you should split it into two different functions.

Reply:
Thanks for clarifying it. I think you are mentioning a different problem with Ethan. For your suggestion, I will need to set an option in .tcl file so that user can input either "% global_place -density ***" or "% global_place -gpu -density ***" to control if they can choose to use CUDA or not. Am I understanding right?

Reply:
#if is a compile time choice so how will you implement the tcl option?

Reply:
Oh! I am not sure if I understood your meaning this time. I can wrap the two code blocks in the pictures as two functions. Then I can use tcl option to control if I will GPU code or not. This means that, if the user wants to use the CPU version with GPU enabled, they just enable GPU in CMake and choose to use CPU code from the .tcl option. Right?

Reply:
![image](https://user-images.githubusercontent.com/57721574/181587623-d0477fd7-071e-4dcc-8416-0f3ccea09113.png)
The structur would be like this?

Reply:
Yes the user should be able to choose gpu or cpu in any case.  If they choose gpu and it wasn't enabled then  give a warning and automatically switch to CPU.

Reply:
This is a good point. I should add this as well. Thanks!

Reply:
Hi Matt, I submitted a new version, in which I modified the CMakelists and added a flag "-force_cpu" via the .tcl command line. 
In this way, if GPU is enabled by CMake, the program uses CUDA solver by default and uses CPU solver if the "force_cpu" is set. If there is no available GPU, it turns into the CPU solver. 
If GPU is not enabled, the program uses CPU solver all the time.
![image](https://user-images.githubusercontent.com/57721574/182052899-ce1bb1c4-50b8-49b2-a7f8-de24d9ee0511.png)